File: cloud-run-bot/tests/paylas_kazan.test.js
const assert = require('assert');
const allowlist = require('../domain_allowlist');
const linkScraperService = require('../link_scraper_service');

async function runTests() {
  console.log('🚀 Starting Teknosa Paylaş Kazan Unit & Integration Tests...\n');

  const paylasKazanUrl = 'https://paylaskazan.teknosa.com/teknosa-F8NSB38NC3';

  // 1. Yönlendirme Çözümü (Redirect Resolution)
  console.log('--- 1. Yönlendirme Çözümü ---');
  const resolvedUrl = await linkScraperService.resolveUrlRedirects(paylasKazanUrl);
  console.log(`Resolved URL: ${resolvedUrl}`);
  assert.ok(resolvedUrl.includes('teknosa.com'), 'Çözülen URL teknosa.com içermeli');
  assert.ok(resolvedUrl.includes('-p-790182989'), 'Çözülen URL ürün kodunu (-p-790182989) içermeli');
  assert.ok(resolvedUrl.includes('shopId=2442'), 'Çözülen URL satıcı kimliğini (shopId=2442) korumalı');

  // 2. Allowlist & Domain Doğrulaması (Çözümlenen Nihai Link Üzerinden)
  console.log('\n--- 2. Çözümlenen URL Allowlist Doğrulaması ---');
  const domainAllowed = allowlist.isDomainAllowed(resolvedUrl);
  console.log(`isDomainAllowed(resolvedUrl): ${domainAllowed}`);
  assert.strictEqual(domainAllowed, true, 'Çözülen URL teknosa.com allowlist tarafından onaylanmalı');

  const isProduct = allowlist.isProductUrl(resolvedUrl);
  console.log(`isProductUrl(resolvedUrl): ${isProduct}`);
  assert.strictEqual(isProduct, true, 'Çözülen URL -p-... regex kuralı ile eşleşmeli');

  const storeKey = allowlist.getStoreKeyForUrl(resolvedUrl);
  console.log(`getStoreKeyForUrl(resolvedUrl): ${storeKey}`);
  assert.strictEqual(storeKey, 'teknosa', 'Mağaza anahtarı "teknosa" olmalı');

  // 3. Uçtan Uca Scraping Doğrulaması
  console.log('\n--- 3. Uçtan Uca Scraping Doğrulaması ---');
  const data = await linkScraperService.scrapeProductFromUrl(paylasKazanUrl);

  console.log('Scraped Title:        ', data.title);
  console.log('Scraped Price:        ', data.price, 'TL');
  console.log('Scraped OriginalPrice:', data.originalPrice, 'TL');
  console.log('Scraped Image:        ', data.imageUrl);
  console.log('Scraped Brand:        ', data.brand);

  assert.ok(data.title && data.title.includes('iPhone XR'), 'Başlık iPhone XR içermeli');
  if (data.price !== null && data.price !== undefined) {
    assert.ok(Number(data.price) > 0, 'İndirimli fiyat pozitif olmalı');
    if (data.originalPrice !== null && data.originalPrice !== undefined) {
      assert.ok(Number(data.originalPrice) >= Number(data.price), 'İndirimsiz fiyat >= indirimli fiyat olmalı');
    }
  } else {
    console.log('ℹ️ Canlı ürün şu an stokta tükenmiş (price=null), başlık/görsel/marka başarıyla doğrulandı.');
  }
  assert.strictEqual(data.brand, 'Apple', 'Marka Apple olmalı');
  assert.ok(data.imageUrl && data.imageUrl.startsWith('http'), 'Geçerli bir görsel URL bulunmalı');

  console.log('\n🎉 ALL TEKNOSA PAYLAŞ KAZAN TESTS PASSED SUCCESSFULLY!\n');
}

runTests().catch(err => {
  console.error('❌ Test failed with error:', err);
  process.exit(1);
});

File: test/paylas_kazan_test.dart
// ignore_for_file: avoid_print
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:sicak_firsatlar/services/domain_allowlist_service.dart';
import 'package:sicak_firsatlar/services/link_preview_service.dart';

void main() {
  HttpOverrides.global = null;

  group('Teknosa Paylaş Kazan Tests', () {
    const paylasKazanUrl = 'https://paylaskazan.teknosa.com/teknosa-F8NSB38NC3';

    test('validateUrl should validate Paylaş Kazan URL end-to-end via shortlink resolution', () async {
      final validationResult = await DomainAllowlistService.validateUrl(paylasKazanUrl);
      expect(validationResult, equals(UrlValidationResult.valid));
    });

    test('Allowlist and Domain validation tests for resolved canonical URL', () async {
      final linkService = LinkPreviewService();
      final resolved = await linkService.resolveUrlRedirects(paylasKazanUrl);

      expect(DomainAllowlistService.isDomainAllowed(resolved), isTrue);
      expect(DomainAllowlistService.isProductUrl(resolved), isTrue);
      expect(DomainAllowlistService.getStoreNameForUrl(resolved), equals('teknosa'));
    });

    test('resolveTeknosaPaylasKazan should extract canonical product URL with shopId', () async {
      final linkService = LinkPreviewService();
      final resolved = await linkService.resolveTeknosaPaylasKazan(paylasKazanUrl);

      print('Resolved Paylaş Kazan URL: $resolved');
      expect(resolved, contains('teknosa.com'));
      expect(resolved, contains('-p-790182989'));
      expect(resolved, contains('shopId=2442'));
    });

    test('resolveUrlRedirects should route and resolve Paylaş Kazan link', () async {
      final linkService = LinkPreviewService();
      final resolved = await linkService.resolveUrlRedirects(paylasKazanUrl);

      print('Resolved through resolveUrlRedirects: $resolved');
      expect(resolved, contains('teknosa.com'));
      expect(resolved, contains('-p-790182989'));
      expect(resolved, contains('shopId=2442'));
    });

    test('validateUrl should validate Paylaş Kazan URL end-to-end', () async {
      final validationResult = await DomainAllowlistService.validateUrl(paylasKazanUrl);
      expect(validationResult, equals(UrlValidationResult.valid));
    });

    test('fetchMetadata should scrape product details from Paylaş Kazan link', () async {
      final linkService = LinkPreviewService();
      final result = await linkService.fetchMetadata(paylasKazanUrl);

      expect(result, isNotNull);
      print('Scraped Title:         ${result!.title}');
      print('Scraped Price:         ${result.price}');
      print('Scraped OriginalPrice: ${result.originalPrice}');
      print('Scraped Image:         ${result.imageUrl}');
      print('Scraped Brand:         ${result.brand}');

      expect(result.title, contains('iPhone XR'));
      if (result.price != null) {
        expect(result.price, isPositive);
      }
      expect(result.brand, equals('Apple'));
    });
  });
}
